### PR TITLE
Streamline HTML generation and consolidate templates

### PR DESCRIPTION
**Pull request by steverep**
## Proposed change

_**Note:** The runtime chunk changes have been removed due to issues with Webpack.  See discussion below._

The main goal here is to separate Webpack's runtime code into its own chunk per [Webpack's caching guide](https://webpack.js.org/guides/caching/#extracting-boilerplate).  Currently that code is bundled in with each entrypoint's initial chunk.  Since it includes an internal manifest that changes each version, the large entry chunks can never be long-term cached.  This fixes that.

In addition, since the index page of the main app includes 2 entrypoints, the runtime is also being included and instantiated twice there.  This fixes that too with a single runtime shared amongst all entrypoints.

Now the fun part... the `entry-html` tasks and templates already have quite a bit of repetitive code, so rather than make that much worse, I completely overhauled those tasks and tried to consolidate some template HTML.  Here's the rough summary:

- Two main functions (dev and prod) now control the tasks.  Each project just gives it the proper paths and a map of pages to entrypoint(s).  This also means that several that were not being minified now are.
- Nested templates are now passed the same data and have relative paths so they can work anywhere
- Consolidated the ES5 loading into a child template used by many pages.  The latest loads were too much of a hodgepodge so maybe next time.
- Got rid of supervisor function writing code and just made it a template

There's certainly more that can be streamlined here, but this is a good start I think.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by steverep** [comment]
Had to disable the runtime chunk manually for the mocha tests to avoid an error.  It's caused by instant-mocha overriding the output filename to `"main.js"`.  The runtime chunk is irrelevant for unit tests anyway.

Ref: https://github.com/privatenumber/instant-mocha/issues/47

**Comment by balloob** [comment]
The `core.ts` entrypoint needs to load as fast as possible and not rely on other files in the process. It's purpose is to start the loading of data while the browser is busy processing the JavaScript that builds up the frontend. Therefore, it should not have its webpack runtime extracted.

Another thing to note is that Home Assistant is served in two ways: locally without HTTPS and remotely with HTTPS. When it's local, the connection speed is fast. Caching might not be super important. When it's remote, connection speeds are often slow, but caching is managed by the service worker which pre-downloads all the files locally.

**Comment by steverep** [comment]
> The `core.ts` entrypoint needs to load as fast as possible and not rely on other files in the process. It's purpose is to start the loading of data while the browser is busy processing the JavaScript that builds up the frontend. Therefore, it should not have its webpack runtime extracted.

For modern builds, they are both loaded async, so core is not waiting for anything, right?  The only difference is the extra file.  Plus, core is dependent on the runtime to load fully anyway because it controls the loading of any split chunks.

> Another thing to note is that Home Assistant is served in two ways: locally without HTTPS and remotely with HTTPS. When it's local, the connection speed is fast. Caching might not be super important. When it's remote, connection speeds are often slow, but caching is managed by the service worker which pre-downloads all the files locally and updates in the background.

Maybe I'm missing your point here.  The goal is to allow the service worker to cache the large entry/initial chunks longer.  Currently they get busted anytime a new frontend is released even though most of the time the actual changes are much deeper in the tree.

File: test/webpack.config.js [hunk 1]
```diff
@@ -1,8 +1,14 @@
 import webpack from "../build-scripts/webpack.cjs";
 
-export default webpack.createAppConfig({
+const config = webpack.createAppConfig({
   isProdBuild: false,
   latestBuild: true,
   isStatsBuild: false,
-  isTestBuild: true,
+  isTestBuild: false,
```
[the remark this answers is not in the document]
Comment (steverep) [reply test/webpack.config.js:7]: Just to turn on source maps so any test failures have a useful stack trace.  I thought I needed it to debug the error at first.  Unrelated, but correct I think.

**Comment by bramkragten** [comment]
Core should (ideally) not have any split chunks. It should all be in the entrypoint

**Comment by steverep** [comment]
> Core should (ideally) not have any split chunks. It should all be in the entrypoint

Just checked my production build and that looks to be correct.  It stands alone. 👍🏻  However, it still depends on the runtime code because of all the `__webpack*` that replaces the imports and exports.

Still, I'm failing to see where any measurable slow down would occur as a result of this change.  There might even be a slight speed up under normal circumstances, and certainly should be faster after an update or when coming from the login or onboarding pages.

**Comment by balloob** [comment]
It loads and starts parsing `core.ts`, then right away realizes it needs to load another chunk and needs to initiate another request and wait for that to resolve.

Maybe the caching in the server worker has changed, but I was under the impression that it would throw away all the cache of the old service worker version and download a fresh copy.

**Comment by steverep** [comment]
> It loads and starts parsing `core.ts`, then right away realizes it needs to load another chunk and needs to initiate another request and wait for that to resolve.

I think maybe you're misunderstanding?  Core doesn't load the runtime.  The runtime chunk is an initial chunk just like the other entrypoints.  It gets preloaded and dynamically imported asynchronously alongside the others from the HTML (hence the bulky changes).  In fact, you could await the core and app loads before importing the runtime, but I don't see much point there.

So any performance loss would come from just the overhead of dealing with 1 extra file and, absent cache, 1 extra HTTP request.  But for the index page, that gets offset by the facts that:
- It's preloaded while the page is parsed
- About 270 lines of repeated Webpack code is dropped (double that if the custom-panel entry also gets loaded)
- If navigating from the login page, it's already in memory



> Maybe the caching in the server worker has changed, but I was under the impression that it would throw away all the cache of the old service worker version and download a fresh copy.

I assume you're referring to the [`install` event handler](https://github.com/home-assistant/frontend/blob/f4fa0ceef520eab757b3e5f13df84f1b56f209f7/src/entrypoints/service_worker.ts#L181)?  That only deletes the "runtime cache", which does not include everything in the precache manifest.

**Comment by steverep** [comment]
Perhaps this will help.  Here's a screenshot of the initiator chain for core from a network recording in Chrome dev tools.  Note it's only dependent request is to the web socket API:

screenshot  of initiator chain for core.js chunk

On the other hand, the app entry needs the runtime to request all the async chunks:

screenshot  of initiator chain for runtime.js chunk

_Please forgive my total lack of precision snipping screenshots!_ 👨🏻‍🦯

**Comment by bramkragten** [comment]
Splitting the runtime in a single chunk also means we are loading the runtime code for other entrypoints (onboarding, auth, custom panel) while we dont need them in the app? This would increase the code we are loading? Meaning the runtime code of just the core entrypoint is way smaller than the runtime code for all the other entrypoints combined that we now need?

**Comment by steverep** [comment]
> Splitting the runtime in a single chunk also means we are loading the runtime code for other entrypoints (onboarding, auth, custom panel) while we dont need them in the app? This would increase the code (manifests) we are loading?

Yes, that's true for the manifest part of the runtime, but offset by not redefining things like `function __webpack_require_` and other code multiple times for each entry.  There's also a fair amount of repeats in the manifest since it shares chunks with authorize and onboarding.  Custom panel doesn't load any other chunks.

> Meaning the runtime code of just the core entrypoint is way smaller than the runtime code for all the other entrypoints combined that we now need?

The vast majority of it is already needed because app loads alongside core, so I don't think it's fair to look at the size of its runtime only.

And now to confuse myself... Here are the sizes in KB I get from back to back builds:

| Entry | Current | New |
|:---- |:----:|:----:|
| runtime | n/a | 16 |
| app | 311 | 297 |
| core | 19 | 21 |
| custom-panel | 35 | 39 |
| service_worker | 67 | 67 |
| authorize | 382 | 383 |
| onboarding | 413 | 406 |



The extra couple KB being picked up by some of the entries is baffling me so I'll investigate that.

**Comment by balloob** [comment]
The page now needs to parse almost twice as much JavaScript (16KB runtime + core is 2KB bigger) before it can do its pre-fetch:

- access local storage to check for existing tokens
- refresh the tokens
- open a websocket connection
- start pre-fetching data needed for the UI. 

And given that most use cases are to show the dashboard etc, current core + app is 330KB, with runtime split out it's 334KB.

I would assume that browsers cache the runtime JS partially parsed. This can make loading that part faster. However, it will always be more work than the current situation, where core.ts doesn't rely on it.

**Comment by steverep** [comment]
@balloob you're certainly correct - there's more work with the single runtime just for core to arrive at opening the web socket.  Given the small sizes and async nature, I think real world timing differences would be immeasureable, but still, no need to make it worse. 👍🏻 

The size increases in the table above were really bugging me though, so I ran another build with webpack set to give a separate runtime chunk for each entry:

| Entry | Main JS | Runtime |
|:---- |:----:|:----:|
| app | 296K | 16K |
| core | 17K | 2.6K |
| custom-panel | 33K | 2.7K |
| service_worker | 68K | 995 |
| authorize | 376K | 6.8K |
| onboarding | 403K | 11K |

Now those numbers make good sense as they just each basically add up to the current entry sizes.

After digging in a little more, I'm realizing my mistake was not properly factoring in tree shaking and minimization.  When an entry has only its one initial chunk, as is the case for core, custom-panel, and service_worker, Webpack and Terser are able to drop runtime code and compress them significantly more.  Not to mention, if there's only the initial chunk, then there's no caching benefit to separating the runtime anyway.  It'd be nice if Webpack realized that automatically or at least gave a warning.

So I think the solution here to get the caching benefits is to separate the runtime for app, authorize, and onboarding, and leave the others alone.  I'll need to refactor a bit, but that's pretty easy.

**Comment by steverep** [comment]
Any strong opinions on whether app, authorize, and onboarding should share the runtime or each have their own?  Multiple is better for caching and sharing produces a slightly smaller bundle.

**Comment by bramkragten** [comment]
I think it is fine to make those share the runtime

**Comment by balloob** [comment]
I don't think it matters?

People see onboarding once in the lifetime of their HA install. Authorization is also seen once per device.

**Comment by steverep** [comment]
Yeah, I didn't think it mattered much either and my kids actually stayed in bed last night, so I had already implemented the one per page strategy.  I just need to test everything again.  I'll push it later.

**Comment by steverep** [comment]
Ugh...

```
TypeError: __webpack_require__.a is not a function
```

This is global namespace pollution at its finest. That's the async chunk loader.  App runtime defines it.  Core's internal runtime doesn't.

File: test/webpack.config.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (steverep) [reply test/webpack.config.js]: The fix I provided to `instant-mocha` was already published and merged in #16160, so I just reverted this commit anyway.

**Comment by steverep** [comment]
It looks like the only way forward here regarding the runtime chunk would be to have each entry with it's own runtime.  As soon as core or custom-panel get imported and overwrite `__webpack_require__`, app will always fail.  And it seems silly to split them into 2 requests for no benefit.  I'll have to file an issue with Webpack.

In the meantime, I'll rebrand this just for the streamlining of the HTML generation, and I've got other improvements in mind that may clash with further testing here anyway.

**Comment by steverep** [comment]
Alright, removed the runtime chunk stuff for now.  Hopefully we can revisit soon.

**Review by bramkragten — APPROVED** [review]

**Pull request merged**